### PR TITLE
Add batch map-reduce compute

### PR DESCRIPTION
**Pull request by sadikovi**
This PR adds batch computation of tasks -which can be useful (certainly is in my use case) when training model which is an individual task in Spark. This allows you to schedule as few partitions as possible so each task will get as much resources as needed.

PR introduces `BatchMapRDD` and `BatchReduceRDD`. The main idea is to trade memory limitations by increasing time of computation for each batch. Specifically we do not try scheduling entire pool of tasks, but split them into batches to run only after previous batch is finished. `BatchMapRDD` allows to construct this graph by passing dependency, though parent is the same for all batch map-side RDDs. `BatchReduceRDD` will take root of the graph and traverse to collect dependencies and ensure that all original partitions are mapped. While computing it collects shuffle output for each partition and drops partition index from values.

You can use it like this:
```
import org.hammerlab.magic.rdd.batch.implicits._
val rdd = sc.parallelize(Seq("a", "b", "c", "d", "e", "f", "g", "h"), 10)
val res = rdd.batch(numPartitionsPerBatch = 4)
res.collect

// or like this:
val rdd = sc.parallelize(Seq((1, true), (2, false), (3, true)), 8)
val res = rdd.batch(3)
res.collect
```

RDD operation graph in the first case will look like this:
```
(10) BatchReduceRDD[9] at RDD at batches.scala:134 []
 +-(2) BatchMapRDD[8] at RDD at batches.scala:74 []
    +-(4) BatchMapRDD[7] at RDD at batches.scala:74 []
       +-(4) BatchMapRDD[6] at RDD at batches.scala:74 []
          |  ParallelCollectionRDD[5] at parallelize at <console>:30 []
```
Each `BatchMapRDD` corresponds to one of the batches 4, 4, 2. In Spark UI you will see separate stages launched for each batch.

screenshot from 2017-02-09 14 01 56

### CODE REVIEW

**Comment by sadikovi** [comment]
@ryan-williams can you review and give a feedback on overall feature? Thanks.

**Comment by ryan-williams** [comment]
This looks pretty cool @sadikovi; I sent you https://github.com/sadikovi/magic-rdds/pull/5 with a file-restructuring proposal, lmk what you think about that.

If I understand the feature itself correctly, you want individual tasks to be able to use all the memory on an executor (sort of like having `spark.executor.cores=1` for a specific stage, so you do that by running waves of jobs, each with a set number of tasks (which is presumably equal to the number of executors)? Do I have that right?

**Comment by sadikovi** [comment]
@ryan-williams Yep, just had a look, looks great!

Yes, your understanding is correct, in some cases one would rather have more resources allocated per task rather than parallelizing as many tasks as possible. One use case is when I run algorithm in Spark and I need to compute small matrices, this allows me to parallelize them and have more memory for computation (I have some code that does size-to-executor-memory estimation). Also this is better than running multiple jobs for each batch of matrices and collecting result on a driver.

Currently I do not hardcode batch size as number of executors, there is parameter to pass `numPartitionsPerBatch`, so it is up for engineer to find optimal size they can run.

**Comment by sadikovi** [comment]
Let me run couple of manual tests on larger data.

**Comment by sadikovi** [comment]
Looks like it works with chaining of `repartition -> map -> batch -> repartition -> collect`:)

**Comment by sadikovi** [comment]
~~DAG visualization does not show the same output as `toDebugString`. I am not sure if we need to address that.~~ All good, they are all shuffle boundaries.

**Comment by ryan-williams** [comment]
OK, I'm basically fine to merge this, though I'd like to maybe document it a little more/differently, which I can help with.

When you said

> rather than parallelizing as many tasks as possible

just to reiterate, the "as many … as possible" here is actually "`spark.executor.cores` tasks concurrently on each executor", and the whole point of this feature is to give you a (somewhat indirect/clumsy) lever for executing a "job" while forcing fewer than `spark.executor.cores` tasks to run concurrently on each executor?

If Spark allowed for tweaking [the maximum number of concurrent tasks per executor] in a per-job fashion, that would solve this problem for you more cleanly, yes?

**Comment by sadikovi** [comment]
Well, option `spark.executor.cores` did not work for me, when I tried it, plus, this forces that many cores for a duration of entire application.

Yes, you are right. If I can tell Spark that at this particular step I want to run different number of tasks simultaneously, this would totally solve the problem, but it does not seem to be a way to do it.

My algorithm works as follows:
- Take input from Parquet, run bunch of SQL queries with aggregations and everything to do some intelligent split into matrices. This step requires as much parallelism as possible, and Spark in default configuration suits just fine.
- Next step is converting each partition (as set of rows) into matrix MxN. Do some computation and return result that is smaller than input matrix.
- Then I run some SQL queries again to store results into a nice table.

Let's say, I have 4 executors (4 cores and 40G of memory each) and 16 matrices. If each matrix is less than 10G, I can execute matrix per core, it is small enough to run without any problems. So Spark with schedule all 16 tasks, and run 4 matrices per executor at a time. Then I get larger data that requires matrices to be 10+G. Now I need to execute less matrices per executor (per JVM), otherwise I might get OOM of running 4 10+G matrices on 40G JVM. With `executor.cores` option, I would have to tell that each task acquires 2+ cores, meaning it affects previous steps, not just matrix. With this approach, it allows me to just run matrices in batches, without affecting other parts.

Again, this is not a super efficient way of doing it, considering that it does not give control over scheduling per executor, just says how many tasks are allowed to run at a time. In the worst case you would still schedule all matrices on the same executor, but in practice it is not the case - it usually works as task per executor then.

As I mentioned previously, I can just run many RDDs in batches, but then I will have to collect results on a driver and then parallelize into RDD again. This is how it worked previously, and, once job finishes, driver program takes very long time to respond. With this, it is like doing normal shuffle.

**Comment by sadikovi** [comment]
It might sound like very inefficient way of using Spark, but it works and is fun to do:)
Plus, I did not know you could do this sort of thing in Spark before:)

**Comment by ryan-williams** [comment]
Cool, that all makes sense, I wasn't saying you could do this with `spark.executor.cores`, just clarifying that the behavior you're simulating here is similar to what one might have wanted from a (currently non-existent) per-job-`spark.executor.cores`-parameter.

Do you have any interest in recording the need for this somewhere upstream, e.g. to the spark-dev list or on a Spark JIRA? I'm sort of ambivalent about doing so, but maybe it would be good to do just for the record…

Also I think we should probably put a note on the README about this functionality? It can either include a complete description, or just link to the README that I dropped in the `batch` package.

You could write that and then I might do an editing pass, or I can just write something based on what you've said here. Let me know what you prefer.

**Comment by sadikovi** [comment]
Yep, will update README.

File: README.md [hunk 1]
```diff
@@ -163,7 +163,7 @@ Adds `.lazyZipWithIndex`, which is functionally equivalent to [`RDD.zipWithIndex
 
 #### [SlidingRDD](https://github.com/hammerlab/magic-rdds/blob/master/src/main/scala/org/hammerlab/magic/rdd/sliding/SlidingRDD.scala)
 
-Exposes `.sliding` methods (and several variants) in the spirit of [Scala collections' similar API](https://github.com/scala/scala/blob/v2.10.5/src/library/scala/collection/IterableLike.scala#L164):
```
Comment (sadikovi) [inline README.md:166]: My editor just removes whitespaces at the end of line.

**Comment by sadikovi** [comment]
@ryan-williams I updated both READMEs, can you have a look? Thanks!

**Comment by sadikovi** [comment]
@ryan-williams thanks for merging this!

**Comment by ryan-williams** [comment]
Thanks for the feature! I just [released 1.4.1 to Maven](https://oss.sonatype.org/content/repositories/releases/org/hammerlab/magic-rdds_2.11/1.4.1/) from HEAD :)

**Pull request merged**